### PR TITLE
Optimise Security Policy creation for non-admin users

### LINKED ISSUE #473: User security policy is locked for too long when Mauro catalogue is large
# Description

I created a non-administrator user on the NHS Data Dictionary test instance of Mauro, this contains 7 versioned folders representing data dictionary branches/releases, so there is a lot of content.

This regular user has trouble logging in, specifically calling the endpoint `GET /session/isApplicationAdministration`. This will build a user security policy, but it takes too long, over 60 seconds locked, then fails with a 500 response.

This does not happen when signing in as an administrator, I have not confirmed yet if an existing user can still sign in who was created before all the content was imported to the catalogue.

# Steps to reproduce

This is reproducable on the NHS Data Dictionary test instance of Mauro at https://ec2-13-42-94-180.eu-west-2.compute.amazonaws.com/, but not everyone will have access to this site due to IP restrictions. However:

1. Go to MDM UI
2. Sign in as "reader@systemc.com". Ask for the password from @pjmonks 
3. Click the button to continue.

# Expected behavior

The user should sign in to Mauro as expected in a timely manner.

# Error details

{
    "status": 500,
    "reason": "Internal Server Error",
    "errorCode": "USP",
    "message": "UserSecurityPolicy has been locked for more than 60 seconds",
    "path": "/api/session/isApplicationAdministration",
    "version": "5.4.0-SNAPSHOT",
    "exception": {
        "type": "ApiInternalException",
        "message": "UserSecurityPolicy has been locked for more than 60 seconds",
        "stacktrace": [
            "uk.ac.ox.softeng.maurodatamapper.security.policy.UserSecurityPolicy.holdForLock(UserSecurityPolicy.groovy:311)",
            "uk.ac.ox.softeng.maurodatamapper.security.policy.UserSecurityPolicy.getApplicationPermittedRoles(UserSecurityPolicy.groovy:90)",
            "uk.ac.ox.softeng.maurodatamapper.security.policy.GroupBasedUserSecurityPolicyManager.hasApplicationLevelRole(GroupBasedUserSecurityPolicyManager.groovy:776)",
            "uk.ac.ox.softeng.maurodatamapper.security.policy.GroupBasedUserSecurityPolicyManager.isApplicationAdministrator(GroupBasedUserSecurityPolicyManager.groovy:553)",
            "uk.ac.ox.softeng.maurodatamapper.core.session.SessionController.isApplicationAdministrationSession(SessionController.groovy:38)",
            "org.grails.core.DefaultGrailsControllerClass$MethodHandleInvoker.invoke(DefaultGrailsControllerClass.java:223)",
            "org.grails.core.DefaultGrailsControllerClass.invoke(DefaultGrailsControllerClass.java:188)",
            "org.grails.web.mapping.mvc.UrlMappingsInfoHandlerAdapter.handle(UrlMappingsInfoHandlerAdapter.groovy:90)",
            "org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1071)",
            "org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:964)",
            "org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)",
            "org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)",
            "org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)",
            "org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:77)",
            "org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:67)",
            "org.springframework.boot.web.servlet.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:126)",
            "org.springframework.boot.web.servlet.support.ErrorPageFilter.access$000(ErrorPageFilter.java:64)",
            "org.springframework.boot.web.servlet.support.ErrorPageFilter$1.doFilterInternal(ErrorPageFilter.java:101)",
            "org.springframework.boot.web.servlet.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:119)"
        ]
    }
}

# Environment

Happens on https://ec2-13-42-94-180.eu-west-2.compute.amazonaws.com/

### PR DESCRIPTION
**Pull request by jamesrwelch**
Fixes #473

### CODE REVIEW

**Comment by jamesrwelch** [comment]
Sorry, the branch was incorrectly named.  But all the tests are passing on it

**Comment by pjmonks** [comment]
I carried out some performance testing to compare sign in times. These were taken to see how long the `GET /session/isApplicationAdministration` endpoint completes in various environments - those marked in bold are the improvements to focus on:

# Administrator user

- NHSE Mauro Test environment - 6.99s
- Local (without fix) - 9.07s
- Local (with fix) - **8.11s**

# Regular user

- NHSE Mauro Test environment - 1 min+ (500 status response)
- Local (without fix) - 1 min+ (500 status response)
- Local (with fix) - **9.54s**

Even if a regular user security policy check takes around 1 - 2 seconds longer than an administrator, this still looks well within tolerance of the 60 second timeout.

File: mdm-security/src/integration-test/groovy/uk/ac/ox/softeng/maurodatamapper/security/CatalogueUserFunctionalSpec.groovy [hunk 1]
```diff
@@ -136,14 +136,16 @@ class CatalogueUserFunctionalSpec extends BaseFunctionalSpec implements Security
         GroupBasedUserSecurityPolicyManager defaultUserSecurityPolicyManager = applicationContext.getBean(
             MdmCoreGrailsPlugin.DEFAULT_USER_SECURITY_POLICY_MANAGER_BEAN_NAME, GroupBasedUserSecurityPolicyManager)
         defaultUserSecurityPolicyManager.lock()
+        userSecurityPolicyService.calculateFolderModelMap()
```
Comment (pjmonks) [inline mdm-security/src/integration-test/groovy/uk/ac/ox/softeng/maurodatamapper/security/CatalogueUserFunctionalSpec.groovy:139]: Question - what does this line do for the test? The return value isn't used and the same method is called further down anyway. Is something being cached in this call which is used later?

**Comment by pjmonks** [comment]
I tested the performance on another endpoint too, `PUT /folders/{folderId}/readByAuthenticated`. It was raised to me that switching this on/off in the UI seemed very slow to actually complete. Thinking this might be related to the security policy for the users, I checked if there was a performance improvement here too.

Running as an administrator

- Local (without fix) - turn "readable by authenticated" on - 1.9 mins to complete
- Local (with fix) - turn "readable by authenticated" on - **4.64s** to complete

So your changes have benefitted other areas too 👍

**Review by pjmonks — APPROVED** [review]
I've tested locally and seen some definite improvements, nice one 🎉 

The code changes look sensible to me, perhaps @joe-crawford might have more opinions about it than me. I'm happy with everything though.

File: mdm-security/src/integration-test/groovy/uk/ac/ox/softeng/maurodatamapper/security/CatalogueUserFunctionalSpec.groovy (hunk 1 above)
Comment (jamesrwelch) [reply mdm-security/src/integration-test/groovy/uk/ac/ox/softeng/maurodatamapper/security/CatalogueUserFunctionalSpec.groovy:139]: Yeah, you're right, it's unnecessary.  I could remove it

**Comment by joe-crawford** [comment]
Can't see a problem with this and great that the performance is much better so good to merge.

Noted your comment Pete but will leave for now given it's in a test.

**Pull request merged**